### PR TITLE
expression, executor: rewrite built-in func length using new expression evaluation architecture

### PR DESCRIPTION
**Pull request by XuHuaiyu**
rewrite built-in func `length` using new expression evaluation architecture
PTAL @hanfei1991 @shenli @hicqu

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -132,32 +132,29 @@ type lengthFunctionClass struct {
...
+	baseIntBuiltinFunc
 }
 
 // eval evals a builtinLengthSig.
 // See https://dev.mysql.com/doc/refman/5.7/en/string-functions.html
-func (b *builtinLengthSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinLengthSig) evalInt(row []types.Datum) (int64, bool, error) {
+	ctx, sc := b.ctx, b.ctx.GetSessionVars().StmtCtx
+	arg0, err := WrapWithCastAsString(b.args[0], ctx)
```
Comment (hanfei1991) [inline expression/builtin_string.go]: We should not add cast here.

File: expression/builtin_string.go [hunk 2]
```diff
@@ -132,32 +132,29 @@ type lengthFunctionClass struct {
 }
 
 func (c *lengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinLengthSig{newBaseBuiltinFunc(args, ctx)}
+	tp := types.NewFieldType(mysql.TypeLonglong)
+	types.SetBinChsClnFlag(tp)
```
Comment (shenli) [inline expression/builtin_string.go]: Consider field length?

File: expression/builtin_string.go [hunk 3]
```diff
@@ -133,6 +133,7 @@ type lengthFunctionClass struct {
 
 func (c *lengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
 	tp := types.NewFieldType(mysql.TypeLonglong)
+	tp.Flen = 10
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: refer to 
https://github.com/mysql/mysql-server/blob/5.7/sql/item_func.h#L1435

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Comment by XuHuaiyu** [comment]
PTAL @zimulala

**Comment by hanfei1991** [comment]
LGTM

File: expression/builtin_string.go [hunk 4]
```diff
@@ -132,32 +132,29 @@ type lengthFunctionClass struct {
...
+	sig := &builtinLengthSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinLengthSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
 // eval evals a builtinLengthSig.
```
Comment (zimulala) [inline expression/builtin_string.go]: Update this comments.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**